### PR TITLE
Data Integrity test with device unplugged during IO

### PR DESCRIPTION
**Pull request by Deixx**
Signed-off-by: Daniel Madej <daniel.madej@intel.com>

### CODE REVIEW

File: test/functional/tests/data_integrity/test_data_integrity_unplug.py [hunk 1]
```diff
@@ -0,0 +1,313 @@
...
+    global fio_seed, tmp_dir, ram_disk
+    cache_dev = TestRun.disks["cache"]
+    core_dev = TestRun.disks["core"]
+
+    sleep_max_s = timedelta(seconds=10)
+
+    with TestRun.step("Test prepare"):
+        random.seed(TestRun.random_seed)
+        fio_seed = random.randint(0, 2 ** 32)
+        TestRun.LOGGER.info(f"FIO seed: {fio_seed}")
```
Comment (arutk) [inline test/functional/tests/data_integrity/test_data_integrity_unplug.py:63]: Please generate a unique seed for fio - the same value should be used to all fio calls within single test run

File: test/functional/tests/data_integrity/test_data_integrity_unplug.py [hunk 2]
```diff
@@ -0,0 +1,313 @@
...
+        tmp_dir = Directory.create_temp_directory()
+        TestRun.LOGGER.info(f"Temporary directory: {tmp_dir.full_path}")
+        ram_disk = RamDisk.create(Size(1, Unit.GiB), 1)[0]
+
+        data_prefill = os.urandom(total_size.value)
```
Comment (arutk) [inline test/functional/tests/data_integrity/test_data_integrity_unplug.py:68]: suggest to generate fill independently on controller and DUT based of shared seed (seed to be generated randomly in each test run)

Comment (arutk) [reply test/functional/tests/data_integrity/test_data_integrity_unplug.py:68]: please use test framework random number generator capabilities to assure random numbers generation reproducibility

**Review by arutk — CHANGES_REQUESTED** [review]

File: test/functional/tests/data_integrity/test_data_integrity_unplug.py [hunk 3]
```diff
@@ -0,0 +1,313 @@
...
+    global max_pattern_io_seqno
+    global g_seqno
+
+    io_log_path = generate_temporary_file_name(tmp_dir, "iolog").stdout
+    num_io = [smax + 1 for _ in range(num_jobs)]
+
+    fio = prepare_base_fio().target(ram_disk.path)
+    for i in range(len(num_io)):
+        if num_io[i] == 0:
+            continue
```
[the remark this answers is not in the document]
Comment (mmichal10) [reply test/functional/tests/data_integrity/test_data_integrity_unplug.py:239]: Or even at the very beginning of this function?

File: test/functional/tests/data_integrity/test_data_integrity_unplug.py [hunk 4]
```diff
@@ -0,0 +1,313 @@
...
+        # collect IO log for 20% steps than requested maximum to have some headroom
+        gen_pattern(int(max(io_seqno) * 1.2))
+
+    return [[g_seqno[j][b][bisect_right(g_seqno[j][b], io_seqno[j]) - 1] for b in
+             range(job_workset_blocks)] for j in range(num_jobs)]
+
+
+def gen_pattern(smax):
+    global max_pattern_io_seqno
+    global g_seqno
```
[the remark this answers is not in the document]
Comment (Deixx) [reply test/functional/tests/data_integrity/test_data_integrity_unplug.py]: `num_jobs` value is only read and not set inside this function in contrast to the listed values - without `global` keyword setting those values would take effect only within the current scope

File: test/functional/tests/data_integrity/test_data_integrity_unplug.py [hunk 5]
```diff
@@ -0,0 +1,313 @@
...
+# seqno of the last I/O taken into account in g_seqno (same for all jobs)
+max_pattern_io_seqno = -1
+
+
+@pytest.mark.os_dependent
+@pytest.mark.parametrize("cache_mode", CacheMode.with_traits(CacheModeTrait.LazyWrites))
+@pytest.mark.require_disk("cache", DiskTypeSet([DiskType.optane, DiskType.nand]))
+@pytest.mark.require_disk("core", DiskTypeLowerThan("cache"))
+@pytest.mark.asyncio
+async def test_data_integrity_unplug(cache_mode):
```
[the remark this answers is not in the document]
Comment (Deixx) [reply test/functional/tests/data_integrity/test_data_integrity_unplug.py:53]: Each iteration uses a different cache line size - it looks like the quality of the test didn't suffer and reduces test overhead.

[the remark this answers is not in the document]
Comment (Deixx) [reply test/functional/tests/data_integrity/test_data_integrity_unplug.py:51]: not man, but it's done

File: test/functional/tests/data_integrity/test_data_integrity_unplug.py (hunk 2 above)
Comment (Deixx) [reply test/functional/tests/data_integrity/test_data_integrity_unplug.py]: Prefill data repeatability is not crucial. This part was rewritten to generate data on DUT side and count checksums from the generated data

File: test/functional/tests/data_integrity/test_data_integrity_unplug.py [hunk 6]
```diff
@@ -0,0 +1,313 @@
...
+
+
+def write_device(device, data):
+    i = 0
+    part_size = 32 * 1024  # cannot send all data at once due to SSH command length limitation
+    while i < len(data) // part_size:
+        part_data = data[i * part_size: (i + 1) * part_size]
+        part_string = str(part_data).lstrip('b')[1:-1].replace(r"\'", r"'\''")
+        command = f"echo -ne '{part_string}' | dd bs={part_size} " \
+                  f"seek={i} of={device.path}"
```
[the remark this answers is not in the document]
Comment (Deixx) [reply test/functional/tests/data_integrity/test_data_integrity_unplug.py]: added

File: test/functional/tests/data_integrity/test_data_integrity_unplug.py [hunk 7]
```diff
@@ -0,0 +1,313 @@
...
+                cache = casadm.load_cache(cache_dev)
+
+            with TestRun.step("Check data"):
+                csums_actual = read_device_md5s(exported_object.path)
+
+                # The last I/O in each job is interrupted by the unplug. It could have made it
+                # to the medium or not. So the last I/O we expect to actually hit the disk
+                # is 'num_io-2' or 'num_io-1' for each job. Below 'n1_' refers to 'num_io-1'
+                # and 'n2_' refers to 'num_io-2'
+                n2_seqno = [ios[j] - 2 for j in range(len(ios))]
```
[the remark this answers is not in the document]
Comment (Deixx) [reply test/functional/tests/data_integrity/test_data_integrity_unplug.py]: done

File: test/functional/tests/data_integrity/test_data_integrity_unplug.py [hunk 8]
```diff
@@ -0,0 +1,313 @@
...
+    return [[g_seqno[j][b][bisect_right(g_seqno[j][b], io_seqno[j]) - 1] for b in
+             range(job_workset_blocks)] for j in range(num_jobs)]
+
+
+def gen_pattern(smax):
+    global max_pattern_io_seqno
+    global g_seqno
+
+    io_log_path = generate_temporary_file_name(tmp_dir, "iolog").stdout
+    num_io = [smax + 1 for _ in range(num_jobs)]
```
[the remark this answers is not in the document]
Comment (Deixx) [reply test/functional/tests/data_integrity/test_data_integrity_unplug.py]: done

File: test/functional/tests/data_integrity/test_data_integrity_unplug.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (Deixx) [reply test/functional/tests/data_integrity/test_data_integrity_unplug.py]: removed at all

File: test/functional/tests/data_integrity/test_data_integrity_unplug.py [hunk 9]
```diff
@@ -0,0 +1,313 @@
...
+
+    fio = prepare_base_fio().target(ram_disk.path)
+    for i in range(len(num_io)):
```
[the remark this answers is not in the document]
Comment (Deixx) [reply test/functional/tests/data_integrity/test_data_integrity_unplug.py]: done

**Review by arutk — APPROVED** [review]

File: test/functional/tests/data_integrity/test_data_integrity_unplug.py [hunk 10]
```diff
@@ -0,0 +1,321 @@
...
+from test_utils.size import Unit, Size
+
+ram_disk, tmp_dir, fio_seed = None, None, None
+num_jobs = 8
+job_workset_size = Size(1, Unit.MiB)
+block_size = Size(1, Unit.Blocks4096)
+job_workset_blocks = int(job_workset_size / block_size)
+
+total_size = num_jobs * job_workset_size
+total_workset_blocks = int(total_size / block_size)
```
[the remark this answers is not in the document]
Comment (Deixx) [reply test/functional/tests/data_integrity/test_data_integrity_unplug.py]: done

File: test/functional/tests/data_integrity/test_data_integrity_unplug.py [hunk 11]
```diff
@@ -0,0 +1,321 @@
...
+        "oflag=direct"
+    )
+    TestRun.executor.run_expect_success(command)
+
+
+# retval[j][b] is the seqno of last I/O to hit block b within job j workset
+# assuming last_io_seqno[j] is the seqno of last I/O committed by job j
+def get_pattern(last_io_seqno):
+    if max(last_io_seqno) > max_log_seqno:
+        # collect IO log for 20% steps than requested maximum to have some headroom
```
[the remark this answers is not in the document]
Comment (Deixx) [reply test/functional/tests/data_integrity/test_data_integrity_unplug.py]: done

**Review by mmichal10 — APPROVED** [review]

**Review by arutk — APPROVED** [review]

**Pull request merged**